Build failure: xpilot-ng

## Steps To Reproduce

1. Install xpilot-ng via whichever mechanism.

## Build log

```
error: builder for '/nix/store/1j63l9js9080vqj3f7p5048c5mxx7k00-xpilot-ng-4.7.3.drv' failed with exit code 2;
       last 25 log lines:
       > suibotdef.c:360:5: error: implicit declaration of function 'Move_point'; did you mean 'Move_init'? [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wimplicit-function-declaration-Wimplicit-function-declaration8;;]
       >   360 |     Move_point(&mv, &answer);
       >       |     ^~~~~~~~~~
       >       |     Move_init
       > suibotdef.c: In function 'Get_object_proximity':
       > suibotdef.c:408:5: error: 'return' with no value, in function returning non-void [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wno-return-mismatch-Wreturn-mismatch8;;]
       >   408 |     return;
       >       |     ^~~~~~
       > suibotdef.c:382:13: note: declared here
       >   382 | static bool Get_object_proximity(player_t *pl, object_t *shot, double sqmaxdist,int maxtime, object_proximity_t *object_proximity){
       >       |             ^~~~~~~~~~~~~~~~~~~~
       > suibotdef.c: In function 'Robot_attack_player':
       > suibotdef.c:875:25: warning: format '%i' expects argument of type 'int', but argument 2 has type 'double' [8;;https://gcc.gnu.org/onlinedocs/gcc-14.2.0/gcc/Warning-Options.html#index-Wformat-Wformat=8;;]
       >   875 |       printf("no: dir= %i\n",direction);
       >       |                        ~^    ~~~~~~~~~
       >       |                         |    |
       >       |                         int  double
       >       |                        %f
       > make[3]: *** [Makefile:345: suibotdef.o] Error 1
       > make[3]: Leaving directory '/build/xpilot-ng-4.7.3/src/server'
       > make[2]: *** [Makefile:223: all-recursive] Error 1
       > make[2]: Leaving directory '/build/xpilot-ng-4.7.3/src'
       > make[1]: *** [Makefile:287: all-recursive] Error 1
       > make[1]: Leaving directory '/build/xpilot-ng-4.7.3'
       > make: *** [Makefile:207: all] Error 2
       For full logs, run 'nix log /nix/store/1j63l9js9080vqj3f7p5048c5mxx7k00-xpilot-ng-4.7.3.drv'.
```

Full build failure log at https://gist.github.com/nipsy/53da223192a3c1e258ddd5d6f480d531.

## Metadata

```
# nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.12.6, NixOS, 25.05 (Warbler), 25.05.20241219.d70bd19`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.24.11`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/8fwsiv0hd7nw1brkvka0jf1frk3m7qkr-source`
```

## Notify maintainers

@7c6f434c 

---

Note for maintainers: Please tag this issue in your PR.

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

--- a/pkgs/games/xpilot/default.nix
+++ b/pkgs/games/xpilot/default.nix
@@ -32,6 +32,9 @@ stdenv.mkDerivation rec {
     zlib
     libXxf86misc
   ];
+
+  patches = [ ./xpilot-ng-gcc-14-fix.patch ];
+
   meta = with lib; {
     description = "Multiplayer X11 space combat game";
     homepage = "http://xpilot.sf.net/";

--- a/pkgs/games/xpilot/xpilot-ng-gcc-14-fix.patch
+++ b/pkgs/games/xpilot/xpilot-ng-gcc-14-fix.patch
@@ -1,0 +1,168 @@
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/mapedit/proto.h xpilot-ng-4.7.3-fixed/src/mapedit/proto.h
+--- xpilot-ng-4.7.3/src/mapedit/proto.h	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/mapedit/proto.h	2024-12-31 14:25:55.790083767 +0100
+@@ -73,15 +73,15 @@
+ int ResizeWidth(HandlerInfo_t info);
+ int ResizeHeight(HandlerInfo_t info);
+ int OpenPreferencesPopup(HandlerInfo_t info);
+-int OpenMapInfoPopup(void);
+-int OpenRobotsPopup(void);
+-int OpenVisibilityPopup(void);
+-int OpenCannonsPopup(void);
+-int OpenRoundsPopup(void);
+-int OpenInitItemsPopup(void);
+-int OpenMaxItemsPopup(void);
+-int OpenProbsPopup(void);
+-int OpenScoringPopup(void);
++int OpenMapInfoPopup(HandlerInfo_t info);
++int OpenRobotsPopup(HandlerInfo_t info);
++int OpenVisibilityPopup(HandlerInfo_t info);
++int OpenCannonsPopup(HandlerInfo_t info);
++int OpenRoundsPopup(HandlerInfo_t info);
++int OpenInitItemsPopup(HandlerInfo_t info);
++int OpenMaxItemsPopup(HandlerInfo_t info);
++int OpenProbsPopup(HandlerInfo_t info);
++int OpenScoringPopup(HandlerInfo_t info);
+ int ValidateCoordHandler(HandlerInfo_t info);
+ int ShowHoles(HandlerInfo_t info);
+ char MapData(int x, int y);
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/mapedit/tools.c xpilot-ng-4.7.3-fixed/src/mapedit/tools.c
+--- xpilot-ng-4.7.3/src/mapedit/tools.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/mapedit/tools.c	2024-12-31 14:26:29.168082279 +0100
+@@ -747,7 +747,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenMapInfoPopup()
++int OpenMapInfoPopup(HandlerInfo_t info)
+ {
+     Window *temp;
+ 
+@@ -790,7 +790,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenRobotsPopup()
++int OpenRobotsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, robots);
+     return 0;
+@@ -801,7 +801,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenVisibilityPopup()
++int OpenVisibilityPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, visibility);
+     return 0;
+@@ -812,7 +812,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenCannonsPopup()
++int OpenCannonsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, cannons);
+     return 0;
+@@ -823,7 +823,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenRoundsPopup()
++int OpenRoundsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, rounds);
+     return 0;
+@@ -834,7 +834,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenInitItemsPopup()
++int OpenInitItemsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, inititems);
+     return 0;
+@@ -845,7 +845,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenMaxItemsPopup()
++int OpenMaxItemsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, maxitems);
+     return 0;
+@@ -856,7 +856,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenProbsPopup()
++int OpenProbsPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, probs);
+     return 0;
+@@ -867,7 +867,7 @@
+ /* Arguments :                                                             */
+ /* Purpose :                                                               */
+ /***************************************************************************/
+-int OpenScoringPopup()
++int OpenScoringPopup(HandlerInfo_t info)
+ {
+     XMapWindow(display, scoring);
+     return 0;
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/suibotdef.c xpilot-ng-4.7.3-fixed/src/server/suibotdef.c
+--- xpilot-ng-4.7.3/src/server/suibotdef.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/suibotdef.c	2024-12-31 14:07:28.316133143 +0100
+@@ -338,12 +338,6 @@
+ 
+ /* } */
+ 
+-struct collans {
+-    int line;
+-    int point;
+-    clvec_t moved;
+-};
+-
+ static bool Wall_in_between_points(int cx1, int cy1, int cx2, int cy2);
+ static bool Wall_in_between_points(int cx1, int cy1, int cx2, int cy2){ /* Wall between two given points?*/
+   
+@@ -405,7 +399,7 @@
+   /* ignore if there is enough time to deal with this object  later */
+   if((time_until_closest < 0) || (time_until_closest > maxtime))
+     /*option instead of fixed value: options.dodgetime))*/
+-    return;
++    return false;
+ 
+   /* get the square of the distance */
+   sqdistance =
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/walls.c xpilot-ng-4.7.3-fixed/src/server/walls.c
+--- xpilot-ng-4.7.3/src/server/walls.c	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/walls.c	2024-12-31 14:09:46.086127001 +0100
+@@ -54,12 +54,6 @@
+ #error "This code assumes that negative numbers round upwards."
+ #endif
+ 
+-struct collans {
+-    int line;
+-    int point;
+-    clvec_t moved;
+-};
+-
+ struct tl2 {
+     int base;
+     int x;
+diff -r -U3 -x '.*.sw?' xpilot-ng-4.7.3/src/server/walls.h xpilot-ng-4.7.3-fixed/src/server/walls.h
+--- xpilot-ng-4.7.3/src/server/walls.h	2010-03-05 02:03:49.000000000 +0100
++++ xpilot-ng-4.7.3-fixed/src/server/walls.h	2024-12-31 14:04:50.198140193 +0100
+@@ -187,4 +187,11 @@
+ 
+ extern int num_polys, num_pstyles, num_estyles, num_bstyles;
+ 
++struct collans {
++    int line;
++    int point;
++    clvec_t moved;
++};
++
++extern void Move_point(const move_t *move, struct collans *answer);
+ #endif

